subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileHierarchySet.java

```diff
@@ -82,6 +87,34 @@ public FileHierarchySet invalidate(String absolutePath) {
             .orElse(empty());
     }
 
+    @Override
+    public FileHierarchySet empty() {
+        return empty(caseSensitivity);
+    }
+
+    @Override
+    public void visitKnownDirectories(Consumer<File> directoryVisitor) {
+        Set<File> visited = new HashSet<>();
+        rootNode.accept((parentToRoot, node) -> {
+            File file = new File(new File("/"), parentToRoot);
+            node.getSnapshot().ifPresent(snapshot -> {
+                if (snapshot.getType() == FileType.Directory) {
```

Comment:
I suppose for regular files/missing file we want to watch the parent directory, if it exists, right?

Reply:
Good point, we should do that.